File: packages/shared/src/preferences.ts
import { z } from "zod";

export const calendarPreferencesResponseSchema = z.object({
  start_hour: z.number().int().min(0).max(23),
  end_hour: z.number().int().min(1).max(24),
  updated_at: z.string().datetime(),
});

export const calendarPreferencesUpdateRequestSchema = z.object({
  start_hour: z.number().int().min(0).max(23),
  end_hour: z.number().int().min(1).max(24),
  last_known_updated_at: z.string().datetime().optional(),
});

export const aiModeSchema = z.enum(["proposal_only", "auto_apply"]);
export const captureResearchPolicySchema = z.enum(["proposal_only", "auto_if_safe"]);

export const captureProviderKindSchema = z.enum(["mistral", "openai", "heuristic"]);
export const syncChannelSchema = z.enum([
  "in_app",
  "push",
  "email",
  "web",
  "mobile",
  "whatsapp",
  "telegram",
  "instagram_business",
  "mobile_share_extension",
  "browser_extension",
]);
export const syncOutputTypeSchema = z.enum(["sync_response", "alert", "digest", "reminder"]);

export const syncChannelByOutputTypeSchema = z.object({
  sync_response: syncChannelSchema,
  alert: syncChannelSchema,
  digest: syncChannelSchema,
  reminder: syncChannelSchema,
});

export const syncChannelByOutputTypeUpdateSchema = z.object({
  sync_response: syncChannelSchema.optional(),
  alert: syncChannelSchema.optional(),
  digest: syncChannelSchema.optional(),
  reminder: syncChannelSchema.optional(),
});

export const syncChannelPreferencesSchema = z.object({
  preferred_channel: syncChannelSchema,
  available_channels: z.array(syncChannelSchema).min(1),
  channel_by_output_type: syncChannelByOutputTypeSchema,
  input_channel: syncChannelSchema,
  output_channel: syncChannelSchema,
});

export const syncChannelPreferencesUpdateSchema = z.object({
  preferred_channel: syncChannelSchema.optional(),
  available_channels: z.array(syncChannelSchema).optional(),
  channel_by_output_type: syncChannelByOutputTypeUpdateSchema.optional(),
  input_channel: syncChannelSchema.optional(),
  output_channel: syncChannelSchema.optional(),
});

const DEFAULT_SYNC_CHANNEL_PREFERENCES = {
  preferred_channel: "in_app" as const,
  available_channels: ["in_app", "web", "mobile"] as const,
  channel_by_output_type: {
    sync_response: "in_app" as const,
    alert: "in_app" as const,
    digest: "in_app" as const,
    reminder: "in_app" as const,
  },
  input_channel: "in_app" as const,
  output_channel: "in_app" as const,
};

export const captureProviderSettingSchema = z.object({
  provider: captureProviderKindSchema,
  model: z.string().min(1),
  language: z.string().nullable().optional(),
  fallback_enabled: z.boolean().default(true),
});

export const captureProviderPreferencesSchema = z.object({
  transcription: captureProviderSettingSchema.extend({
    language: z.string().default("auto"),
  }),
  ocr: captureProviderSettingSchema,
  vlm: captureProviderSettingSchema,
});

export const aiPreferencesResponseSchema = z.object({
  mode: aiModeSchema,
  auto_apply_threshold: z.number().min(0).max(1),
  max_actions_per_capture: z.number().int().min(1).max(3),
  capture_provider_preferences: captureProviderPreferencesSchema,
  capture_default_agent_id: z.string().uuid().nullable().optional(),
  capture_agent_routing_rules: z.record(z.string(), z.unknown()).default({}),
  editor_default_agent_id: z.string().uuid().nullable().optional(),
  editor_agent_routing_rules: z.record(z.string(), z.unknown()).default({}),
  capture_research_policy: captureResearchPolicySchema.default("proposal_only"),
  sync_model: z.string().default("auto"),
  sync_reasoning_level: z.string().nullable().optional(),
  sync_channel_preferences: syncChannelPreferencesSchema.default({
    preferred_channel: DEFAULT_SYNC_CHANNEL_PREFERENCES.preferred_channel,
    available_channels: [...DEFAULT_SYNC_CHANNEL_PREFERENCES.available_channels],
    channel_by_output_type: {
      ...DEFAULT_SYNC_CHANNEL_PREFERENCES.channel_by_output_type,
    },
    input_channel: DEFAULT_SYNC_CHANNEL_PREFERENCES.input_channel,
    output_channel: DEFAULT_SYNC_CHANNEL_PREFERENCES.output_channel,
  }),
  updated_at: z.string().datetime(),
});

export const aiPreferencesUpdateRequestSchema = z.object({
  mode: aiModeSchema,
  auto_apply_threshold: z.number().min(0).max(1),
  max_actions_per_capture: z.number().int().min(1).max(3),
  capture_provider_preferences: captureProviderPreferencesSchema.optional(),
  capture_default_agent_id: z.string().uuid().nullable().optional(),
  capture_agent_routing_rules: z.record(z.string(), z.unknown()).optional(),
  editor_default_agent_id: z.string().uuid().nullable().optional(),
  editor_agent_routing_rules: z.record(z.string(), z.unknown()).optional(),
  capture_research_policy: captureResearchPolicySchema.optional(),
  sync_model: z.string().optional(),
  sync_reasoning_level: z.string().nullable().optional(),
  sync_channel_preferences: syncChannelPreferencesUpdateSchema.optional(),
  last_known_updated_at: z.string().datetime().optional(),
});

export type CalendarPreferencesResponse = z.infer<
  typeof calendarPreferencesResponseSchema
>;
export type CalendarPreferencesUpdateRequest = z.input<
  typeof calendarPreferencesUpdateRequestSchema
>;
export type AiMode = z.infer<typeof aiModeSchema>;
export type CaptureResearchPolicy = z.infer<typeof captureResearchPolicySchema>;
export type CaptureProviderKind = z.infer<typeof captureProviderKindSchema>;
export type CaptureProviderSetting = z.infer<typeof captureProviderSettingSchema>;
export type CaptureProviderPreferences = z.infer<typeof captureProviderPreferencesSchema>;
export type SyncChannel = z.infer<typeof syncChannelSchema>;
export type SyncOutputType = z.infer<typeof syncOutputTypeSchema>;
export type SyncChannelByOutputType = z.infer<typeof syncChannelByOutputTypeSchema>;
export type SyncChannelPreferences = z.infer<typeof syncChannelPreferencesSchema>;
export type SyncChannelPreferencesUpdate = z.input<typeof syncChannelPreferencesUpdateSchema>;
export type AiPreferencesResponse = z.infer<typeof aiPreferencesResponseSchema>;
export type AiPreferencesUpdateRequest = z.input<
  typeof aiPreferencesUpdateRequestSchema
>;

File: apps/api/src/schemas/preferences.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field
from typing import Literal


class CalendarPreferencesResponse(BaseModel):
    start_hour: int = Field(ge=0, le=23)
    end_hour: int = Field(ge=1, le=24)
    updated_at: datetime


class CalendarPreferencesUpdateRequest(BaseModel):
    start_hour: int = Field(ge=0, le=23)
    end_hour: int = Field(ge=1, le=24)
    last_known_updated_at: datetime | None = None


AiMode = Literal["proposal_only", "auto_apply"]
CaptureResearchPolicy = Literal["proposal_only", "auto_if_safe"]
CaptureProviderKind = Literal["mistral", "openai", "heuristic"]
SyncChannelKind = Literal[
    "in_app",
    "push",
    "email",
    "web",
    "mobile",
    "whatsapp",
    "telegram",
    "instagram_business",
    "mobile_share_extension",
    "browser_extension",
]
SyncOutputType = Literal["sync_response", "alert", "digest", "reminder"]


class CaptureProviderSetting(BaseModel):
    provider: CaptureProviderKind
    model: str = Field(min_length=1)
    language: str | None = None
    fallback_enabled: bool = True


class CaptureProviderPreferences(BaseModel):
    transcription: CaptureProviderSetting
    ocr: CaptureProviderSetting
    vlm: CaptureProviderSetting


class SyncChannelByOutputType(BaseModel):
    sync_response: SyncChannelKind = "in_app"
    alert: SyncChannelKind = "in_app"
    digest: SyncChannelKind = "in_app"
    reminder: SyncChannelKind = "in_app"


class SyncChannelByOutputTypeUpdate(BaseModel):
    sync_response: SyncChannelKind | None = None
    alert: SyncChannelKind | None = None
    digest: SyncChannelKind | None = None
    reminder: SyncChannelKind | None = None


class SyncChannelPreferences(BaseModel):
    preferred_channel: SyncChannelKind = "in_app"
    available_channels: list[SyncChannelKind] = Field(
        default_factory=lambda: ["in_app", "web", "mobile"]
    )
    channel_by_output_type: SyncChannelByOutputType = Field(
        default_factory=SyncChannelByOutputType
    )
    input_channel: SyncChannelKind = "in_app"
    output_channel: SyncChannelKind = "in_app"


class SyncChannelPreferencesUpdate(BaseModel):
    preferred_channel: SyncChannelKind | None = None
    available_channels: list[SyncChannelKind] | None = None
    channel_by_output_type: SyncChannelByOutputTypeUpdate | None = None
    input_channel: SyncChannelKind | None = None
    output_channel: SyncChannelKind | None = None


class AiPreferencesResponse(BaseModel):
    mode: AiMode
    auto_apply_threshold: float = Field(ge=0, le=1)
    max_actions_per_capture: int = Field(ge=1, le=3)
    capture_provider_preferences: CaptureProviderPreferences
    capture_default_agent_id: UUID | None = None
    capture_agent_routing_rules: dict = Field(default_factory=dict)
    editor_default_agent_id: UUID | None = None
    editor_agent_routing_rules: dict = Field(default_factory=dict)
    capture_research_policy: CaptureResearchPolicy = "proposal_only"
    sync_model: str = "auto"
    sync_reasoning_level: str | None = None
    sync_channel_preferences: SyncChannelPreferences = Field(
        default_factory=SyncChannelPreferences
    )
    updated_at: datetime


class AiPreferencesUpdateRequest(BaseModel):
    mode: AiMode
    auto_apply_threshold: float = Field(ge=0, le=1)
    max_actions_per_capture: int = Field(ge=1, le=3)
    capture_provider_preferences: CaptureProviderPreferences | None = None
    capture_default_agent_id: UUID | None = None
    capture_agent_routing_rules: dict | None = None
    editor_default_agent_id: UUID | None = None
    editor_agent_routing_rules: dict | None = None
    capture_research_policy: CaptureResearchPolicy | None = None
    sync_model: str | None = None
    sync_reasoning_level: str | None = None
    sync_channel_preferences: SyncChannelPreferencesUpdate | None = None
    last_known_updated_at: datetime | None = None
